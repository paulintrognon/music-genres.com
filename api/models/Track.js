const Sequelize = require('sequelize');
const sequelize = require('../db.js').sequelize;

const MusicGenre = require('./MusicGenre.js');

const Track = sequelize.define('track', {
  url: {
    type: Sequelize.STRING,
    allowNull: false,
    notEmpty: true,
    len: [5, 200],
  },
});

Track.belongsTo(MusicGenre);

Track.sync();

module.exports = Track;
