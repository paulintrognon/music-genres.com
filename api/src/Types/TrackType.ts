export type TrackType = {
  id: number
  title: string
  playerName: 'youtube'
  playerTrackId: string
}
