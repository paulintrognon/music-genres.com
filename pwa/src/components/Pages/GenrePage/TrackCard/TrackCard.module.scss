.container {
  display: block;
  width: 100%;
  margin-bottom: 2rem;
}

.play {
  display: block;
  padding: 56.25% 0 0;
  background-position: center;
  background-size: cover;
  text-align: center;
  position: relative;
}

.playIconContainer {
  position: absolute;
  left: 0;
  right: 0;
  top: 0;
  bottom: 0;
  display: flex;
  justify-content: center;
  align-items: center;

  svg {
    width: 80px;
    height: 80px;
  }

  svg path:first-child {
    transition: fill 200ms;
    fill: rgba(30, 30, 30, 0.7);
  }
  &:hover svg path:first-child {
    fill: #F44336;
  }
}
