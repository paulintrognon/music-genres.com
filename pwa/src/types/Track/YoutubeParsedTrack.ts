export type YoutubeParsedTrack = {
  playerName: 'youtube'
  playerTrackId: string
  title: string
  description: string
  tags: string[]
  owner: { name: string; id: string }
}
