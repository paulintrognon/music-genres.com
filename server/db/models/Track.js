module.exports = (sequelize, DataTypes) => {
  const Track = sequelize.define('Track', {
    playerName: {
      type: DataTypes.STRING,
      allowNull: false,
      notEmpty: true,
      len: [3, 30],
    },
    playerTrackId: {
      type: DataTypes.STRING,
      allowNull: false,
      notEmpty: true,
      len: [1, 255],
    },
    title: {
      type: DataTypes.STRING,
      allowNull: false,
      notEmpty: true,
      len: [1, 100],
    },
    description: {
      type: DataTypes.TEXT,
      allowNull: false,
      notEmpty: false,
    },
    upvotes: {
      type: DataTypes.INTEGER,
      allowNull: false,
      defaultValue: 0,
    },
  })

  Track.associate = (models) => {
    Track.belongsToMany(models.MusicGenre, { through: models.MusicGenreTrack })
    Track.hasMany(models.Vote)
  }

  return Track
}
