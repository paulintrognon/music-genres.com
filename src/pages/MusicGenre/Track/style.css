.track-box {
  background-color: #FFF;
  display: inline-block;
  margin: 0;
  vertical-align:top;
  width: 33.33%;
}

.track-box.index-2 {
  margin-top: 20px;
}
.track-box.index-3 {
  margin-top: 40px;
}

.track-box .track-thumb {
  display: block;
  width: 100%;
  padding: 56.25% 0 0 0;
  background-color: #999;
  background-position: center;
  background-size: cover;
  border: 1px solid #fff;
  box-sizing: border-box;
  text-align: center;
  position: relative;
}

.track-box .track-thumb .play-button {
  position: absolute;
  top: 0;
  bottom: 0;
  right: 0;
  left: 0;
  background-repeat: no-repeat;
  background-position: center;
  transition: opacity 200ms;
  cursor: pointer;
}

.track-box .track-thumb .play-button.play-button-normal {
  opacity: 1;
}
.track-box .track-thumb:hover .play-button.play-button-normal {
  opacity: 0;
}
.track-box .track-thumb .play-button.play-button-hover {
  opacity: 0;
}
.track-box .track-thumb:hover .play-button.play-button-hover {
  opacity: 1;
}

.track-box .track-information {
  margin-left: 18px;
  margin-right: 50px;
  min-height: 100px;
}

.track-box .track-title {
  margin-top: 37px;
  margin-bottom: 0;
  height: 100px;
  overflow: hidden;
  font-family: 'Montserrat';
  font-weight: unset;
  text-transform: uppercase;
  color: #8300FC;
  font-size: 26px;
  font-weight: 700;
  text-align: left;
}

.track-box .track-votes {
  margin-top: 43px;
  margin-bottom: 37px;
  font-family: Oxygen;
  color: #5bdc77;
  font-size: 22px;
  font-weight: 400;
  text-align: left;
  display: inline-block;
}

.track-box .track-upvote-button {
  display: inline-block;
  transform: skew(-20deg);
  background-color: #62DD7C;
  margin-left: 20px;
  cursor: pointer;
  text-align: center;
  width: 40px;
  transition: background-color 100ms;
}

.track-box .track-upvote-button:hover {
  background-color: #8300FC;
}

.track-box .track-upvote-button span {
  display: inline-block;
  transform: skew(20deg);
}

.track-box .track-upvote-button.has-upvoted {
  cursor: default;
  background-color: #8300FC;
}
