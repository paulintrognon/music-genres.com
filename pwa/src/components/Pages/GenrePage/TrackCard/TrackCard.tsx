import { useDispatch } from 'react-redux'
import { playTrack } from '../../../../redux/player/player.slice'
import { TrackType } from '../../../../types/Track/TrackType'
import PlayIcon from '../../../Shared/Atoms/Icons/PlayIcon/PlayIcon'
import classes from './TrackCard.module.scss'

interface Props {
  track: TrackType
}
const TrackCard: React.FC<Props> = ({ track }) => {
  const dispatch = useDispatch()

  const handleOnClick = (): void => {
    dispatch(playTrack({ track }))
  }

  return (
    <button type="button" title={track.title} className={classes.container} onClick={handleOnClick}>
      <div
        className={classes.play}
        style={{
          backgroundImage: `url("https://img.youtube.com/vi/${track.playerTrackId}/hqdefault.jpg")`,
        }}
      >
        <div className={classes.playIconContainer}>
          <PlayIcon />
        </div>
      </div>
    </button>
  )
}
export default TrackCard
