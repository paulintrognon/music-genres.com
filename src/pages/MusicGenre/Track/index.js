import React from 'react';

import './style.css';

import playImage from './play.png';
import playHoverImage from './play_hover.png';

export default function Track(props) {
  if (!props.track) {
    return null;
  }
  const track = props.track;
  const marginTop = props.index * 70;
  const style = {
    marginTop: `${marginTop}px`,
  };
  return (
    <div className="track-box" style={style}>
      <div className="track-thumb"
        style={{ backgroundImage: `url('https://img.youtube.com/vi/${track.playerTrackId}/hqdefault.jpg')` }}
        title={`Play "${track.title}"`}
        >
          <p className="play-button play-button-normal"
            style={{ backgroundImage: `url('${playImage}')` }}
            >
          </p>
          <p className="play-button play-button-hover"
            style={{ backgroundImage: `url('${playHoverImage}')` }}
            >
          </p>
      </div>
      <div className="track-information">
        <h3 className="track-title" title={track.title}>
          {track.title.length < 35 ? track.title : track.title.substring(0, 35)+'â€¦'}
        </h3>
        <p className="track-votes">
          {track.upvotes} vote{track.upvotes > 1 ? 's' : ''}
        </p>
        <p className={'track-upvote-button' + (track.hasUpvoted ? ' has-upvoted' : '')} onClick={() => props.upvoteTrackHandler(track.id)}>
          <span>+1</span>
        </p>
      </div>
    </div>
  );
}
