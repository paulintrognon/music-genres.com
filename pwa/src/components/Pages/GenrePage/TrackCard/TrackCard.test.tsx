import React from 'react'
import { createRenderer } from 'react-test-renderer/shallow'
import { shallow } from 'enzyme'
import { TrackType } from '../../../../types/Track/TrackType'
import TrackCard from './TrackCard'

// Mocking useDispatch
import { useDispatch } from 'react-redux'
jest.mock('react-redux')
const useDispatchMock = useDispatch as jest.Mock
const dispatchMock = jest.fn()

beforeEach(() => {
  jest.resetAllMocks()
  useDispatchMock.mockReturnValue(dispatchMock)
})

const track: TrackType = {
  id: 0,
  title: 'Super Track',
  playerName: 'youtube',
  playerTrackId: 'abcb123',
}

it('should render the TrackCard', () => {
  const renderer = createRenderer()
  expect(renderer.render(<TrackCard track={track} />)).toMatchSnapshot()
})

it('should call the action on click', () => {
  const component = shallow(<TrackCard track={track} />)
  component.find('button').simulate('click')
  expect(dispatchMock.mock.calls).toEqual([
    [
      {
        payload: { track },
        type: 'player/playTrack',
      },
    ],
  ])
})
