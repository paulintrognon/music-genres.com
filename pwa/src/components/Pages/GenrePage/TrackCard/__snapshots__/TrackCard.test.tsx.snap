// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the TrackCard 1`] = `
<button
  className="container"
  onClick={[Function]}
  title="Super Track"
  type="button"
>
  <div
    className="play"
    style={
      Object {
        "backgroundImage": "url(\\"https://img.youtube.com/vi/abcb123/hqdefault.jpg\\")",
      }
    }
  >
    <div
      className="playIconContainer"
    >
      <PlayIcon />
    </div>
  </div>
</button>
`;
